File: node/cubrid/02_crud.js
// 02_crud.js — CRUD operations with @cubrid/client.
//
// Demonstrates:
// - CREATE TABLE
// - INSERT rows (single and batch)
// - SELECT with filtering
// - UPDATE rows
// - DELETE rows
// - DROP TABLE cleanup

import { createClient } from "@cubrid/client";

const DB_CONFIG = {
  host: "localhost",
  port: 33000,
  database: "testdb",
  user: "dba",
  password: "",
};

async function setupTable(db) {
  await db.query("DROP TABLE IF EXISTS cookbook_users");
  await db.query(`
    CREATE TABLE cookbook_users (
      id    INT AUTO_INCREMENT PRIMARY KEY,
      name  VARCHAR(100) NOT NULL,
      email VARCHAR(200) UNIQUE,
      age   INT DEFAULT 0
    )
  `);
  console.log("✓ Created table 'cookbook_users'");
}

async function insertRows(db) {
  // Single insert with parameterized query
  await db.query(
    "INSERT INTO cookbook_users (name, email, age) VALUES (?, ?, ?)",
    ["Alice", "alice@example.com", 30],
  );

  // Multiple inserts
  const users = [
    ["Bob", "bob@example.com", 25],
    ["Charlie", "charlie@example.com", 35],
    ["Diana", "diana@example.com", 28],
    ["Eve", "eve@example.com", 32],
  ];
  for (const user of users) {
    await db.query(
      "INSERT INTO cookbook_users (name, email, age) VALUES (?, ?, ?)",
      user,
    );
  }
  console.log(`✓ Inserted ${1 + users.length} rows`);
}

async function selectAll(db) {
  const rows = await db.query(
    "SELECT id, name, email, age FROM cookbook_users ORDER BY id",
  );
  console.log(`\nAll users (${rows.length} rows):`);
  console.log(
    `  ${"ID".padStart(3)}  ${"Name".padEnd(12)}  ${"Email".padEnd(25)}  ${"Age".padStart(3)}`,
  );
  console.log(
    `  ${"---".padStart(3)}  ${"----".padEnd(12)}  ${"-----".padEnd(25)}  ${"---".padStart(3)}`,
  );
  for (const row of rows) {
    console.log(
      `  ${String(row.id).padStart(3)}  ${String(row.name).padEnd(12)}  ${String(row.email).padEnd(25)}  ${String(row.age).padStart(3)}`,
    );
  }
}

async function selectFiltered(db) {
  // Filter by age
  const byAge = await db.query(
    "SELECT name, age FROM cookbook_users WHERE age >= ? ORDER BY age DESC",
    [30],
  );
  console.log(`\nUsers age >= 30 (${byAge.length} rows):`);
  for (const row of byAge) {
    console.log(`  ${String(row.name).padEnd(12)}  age=${row.age}`);
  }

  // Filter by name pattern
  const byName = await db.query(
    "SELECT name, email FROM cookbook_users WHERE name LIKE ?",
    ["%li%"],
  );
  console.log(`\nUsers with 'li' in name (${byName.length} rows):`);
  for (const row of byName) {
    console.log(`  ${String(row.name).padEnd(12)}  ${row.email}`);
  }
}

async function updateRows(db) {
  // Update single row
  await db.query(
    "UPDATE cookbook_users SET age = ? WHERE name = ?",
    [31, "Alice"],
  );
  console.log("\n✓ Updated Alice's age to 31");

  // Update multiple rows
  await db.query(
    "UPDATE cookbook_users SET age = age + 1 WHERE age < ?",
    [30],
  );
  console.log("✓ Incremented age for users younger than 30");
}

async function deleteRows(db) {
  await db.query("DELETE FROM cookbook_users WHERE name = ?", ["Eve"]);
  console.log("\n✓ Deleted Eve");
}

async function cleanup(db) {
  await db.query("DROP TABLE IF EXISTS cookbook_users");
  console.log("\n✓ Cleaned up table 'cookbook_users'");
}

async function main() {
  const db = createClient(DB_CONFIG);
  try {
    await setupTable(db);
    await insertRows(db);
    await selectAll(db);
    await selectFiltered(db);
    await updateRows(db);
    await selectAll(db); // Show updated state
    await deleteRows(db);
    await selectAll(db); // Show final state
  } finally {
    await cleanup(db);
    await db.close();
  }
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: python/pycubrid/02_crud.py
"""02_crud.py — CRUD operations with pycubrid.

Demonstrates:
- CREATE TABLE
- INSERT rows
- SELECT with filtering
- UPDATE rows
- DELETE rows
- DROP TABLE cleanup
"""

from __future__ import annotations

import pycubrid

DB_CONFIG = {
    "host": "localhost",
    "port": 33000,
    "database": "testdb",
    "user": "dba",
    "password": "",
}


def get_connection() -> pycubrid.Connection:
    return pycubrid.connect(**DB_CONFIG)


def setup_table(conn: pycubrid.Connection) -> None:
    """Create the example table."""
    cursor = conn.cursor()
    cursor.execute("DROP TABLE IF EXISTS cookbook_users")
    cursor.execute("""
        CREATE TABLE cookbook_users (
            id    INT AUTO_INCREMENT PRIMARY KEY,
            name  VARCHAR(100) NOT NULL,
            email VARCHAR(200) UNIQUE,
            age   INT DEFAULT 0
        )
    """)
    conn.commit()
    cursor.close()
    print("✓ Created table 'cookbook_users'")


def insert_rows(conn: pycubrid.Connection) -> None:
    """INSERT — add rows to the table."""
    cursor = conn.cursor()

    # Single insert with parameters
    cursor.execute(
        "INSERT INTO cookbook_users (name, email, age) VALUES (?, ?, ?)",
        ("Alice", "alice@example.com", 30),
    )

    # Multiple inserts
    users = [
        ("Bob", "bob@example.com", 25),
        ("Charlie", "charlie@example.com", 35),
        ("Diana", "diana@example.com", 28),
        ("Eve", "eve@example.com", 32),
    ]
    cursor.executemany(
        "INSERT INTO cookbook_users (name, email, age) VALUES (?, ?, ?)",
        users,
    )

    conn.commit()
    cursor.close()
    print(f"✓ Inserted {1 + len(users)} rows")


def select_all(conn: pycubrid.Connection) -> None:
    """SELECT — read all rows."""
    cursor = conn.cursor()
    cursor.execute("SELECT id, name, email, age FROM cookbook_users ORDER BY id")

    rows = cursor.fetchall()
    print(f"\nAll users ({len(rows)} rows):")
    print(f"  {'ID':>3s}  {'Name':12s}  {'Email':25s}  {'Age':>3s}")
    print(f"  {'---':>3s}  {'----':12s}  {'-----':25s}  {'---':>3s}")
    for row in rows:
        print(f"  {row[0]:3d}  {row[1]:12s}  {row[2]:25s}  {row[3]:3d}")

    cursor.close()


def select_filtered(conn: pycubrid.Connection) -> None:
    """SELECT with WHERE clause and parameters."""
    cursor = conn.cursor()

    # Filter by age
    cursor.execute(
        "SELECT name, age FROM cookbook_users WHERE age >= ? ORDER BY age DESC",
        (30,),
    )
    rows = cursor.fetchall()
    print(f"\nUsers age >= 30 ({len(rows)} rows):")
    for row in rows:
        print(f"  {row[0]:12s}  age={row[1]}")

    # Filter by name pattern
    cursor.execute(
        "SELECT name, email FROM cookbook_users WHERE name LIKE ?",
        ("%li%",),
    )
    rows = cursor.fetchall()
    print(f"\nUsers with 'li' in name ({len(rows)} rows):")
    for row in rows:
        print(f"  {row[0]:12s}  {row[1]}")

    cursor.close()


def select_with_fetch_methods(conn: pycubrid.Connection) -> None:
    """Demonstrate fetchone, fetchmany, fetchall."""
    cursor = conn.cursor()
    cursor.execute("SELECT name FROM cookbook_users ORDER BY id")

    # fetchone — one row at a time
    first = cursor.fetchone()
    print(f"\nfetchone():  {first[0]}")

    # fetchmany — batch of rows
    batch = cursor.fetchmany(2)
    print(f"fetchmany(2): {[r[0] for r in batch]}")

    # fetchall — remaining rows
    rest = cursor.fetchall()
    print(f"fetchall():  {[r[0] for r in rest]}")

    cursor.close()


def update_rows(conn: pycubrid.Connection) -> None:
    """UPDATE — modify existing rows."""
    cursor = conn.cursor()

    # Update single row
    cursor.execute(
        "UPDATE cookbook_users SET age = ? WHERE name = ?",
        (31, "Alice"),
    )
    print(f"\n✓ Updated Alice's age (rows affected: {cursor.rowcount})")

    # Update multiple rows
    cursor.execute(
        "UPDATE cookbook_users SET age = age + 1 WHERE age < ?",
        (30,),
    )
    print(f"✓ Incremented age for young users (rows affected: {cursor.rowcount})")

    conn.commit()
    cursor.close()


def delete_rows(conn: pycubrid.Connection) -> None:
    """DELETE — remove rows."""
    cursor = conn.cursor()

    cursor.execute("DELETE FROM cookbook_users WHERE name = ?", ("Eve",))
    print(f"\n✓ Deleted Eve (rows affected: {cursor.rowcount})")

    conn.commit()
    cursor.close()


def cleanup(conn: pycubrid.Connection) -> None:
    """Drop the example table."""
    cursor = conn.cursor()
    cursor.execute("DROP TABLE IF EXISTS cookbook_users")
    conn.commit()
    cursor.close()
    print("\n✓ Cleaned up table 'cookbook_users'")


if __name__ == "__main__":
    conn = get_connection()

    try:
        setup_table(conn)
        insert_rows(conn)
        select_all(conn)
        select_filtered(conn)
        select_with_fetch_methods(conn)
        update_rows(conn)
        select_all(conn)  # Show updated state
        delete_rows(conn)
        select_all(conn)  # Show final state
    finally:
        cleanup(conn)
        conn.close()
